node-graph/graph-craft/src/document.rs

```diff
@@ -40,18 +43,80 @@ fn return_true() -> bool {
+	/// This is similar to having the first input be a NodeInput::Network after the graph flattening.
+	///
+	/// ## Example
+	///
+	/// The `CacheNode` is a pass-through node on cache miss, but on cache hit it needs to avoid evaluating the upstream node and instead just return the cached value.
+	///
+	/// First, let's consider what that would look like using the default composition flow if the `CacheNode` just always acted as a pass-through:
+	///
+	/// ```text
+	///              [ComposeNode]         [ComposeNode]
+	/// ┌───────────────┐  |  ┌───────────────┐  |  ┌───────────────┐
+	/// │               │◄─|──┤               │◄─|──┤               │◄─── EVAL (START)
+	/// │       G       │  |  │PassThroughNode│  |  │       F       │
+	/// │               ├──|─►│               ├──|─►│               │───► RESULT (END)
+	/// └───────────────┘  |  └───────────────┘  |  └───────────────┘
+	/// ```
+	///
+	/// Without manual composition, `ComposeNode`s would be automatically inserted (during the graph rewriting step) between nodes where the arrows are drawn in the diagram above.
+	/// This acts like the function call `F(PassThroughNode(G(input)))` when evaluating `F` with some `input`: `F.eval(input)`.
+	/// - The diagram's upper track of arrows represents the flow of building up the call stack:
+	///   since `F` is the output it is encountered first but deferred to its upstream caller `PassThroughNode` and that is once again deferred to its upstream caller `G`.
+	/// - The diagram's lower track of arrows represents the flow of evaluating the call stack:
+	///   `G` is evaluated first, then `PassThroughNode` is evaluated with the result of `G`, and finally `F` is evaluated with the result of `PassThroughNode`.
+	///
+	/// But to make the `CacheNode` work, we need to override the default composition flow so that `G` is not automatically evaluated when the cache is hit. We need to give the `CacheNode` more manual control over the order of execution. So the `CacheNode` opts into manual composition and, instead of deferring to its upstream caller, it consumes the input directly:
+	///
+	/// ```text
+	///                      ┌───────────────┐    ┌───────────────┐
+	///                      │               │◄───┤               │◄─── EVAL (START)
+	///                      │   CacheNode   │    │       F       │
+	///                      │               ├───►│               │───► RESULT (END)
+	/// ┌───────────────┐    ├───────────────┤    └───────────────┘
+	/// │               │◄───┤               │
+	/// │       G       │    │  Cached Data  │
+	/// │               ├───►│               │
+	/// └───────────────┘    └───────────────┘
+	/// ```
+	///
+	/// Now, the call from `F` directly reaches the `CacheNode` and the `CacheNode` can decide whether to call `G.eval(input_from_f)` in the event of a cache miss or just return the cached data in the event of a cache hit.
```

Comment:
@TrueDoctor @0HyperCube I've further extended the explanation by merging in the flowcharts from `enum NodeInput::ShortCircuit` and adding more context. But I want to make sure I haven't written any inaccuracies, so I'd appreciate if you can both proofread this for correctness. Thanks!

Reply:
Besides the two comments this looks pretty good to me